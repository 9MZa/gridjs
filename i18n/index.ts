import esES from './es_ES';
import frFR from './fr_FR';
import itIT from './it_IT';

export {
  esES,
  frFR,
  itIT
};
