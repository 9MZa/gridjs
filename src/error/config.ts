import BaseError from './base';

class ConfigError extends BaseError {}
export default ConfigError;
