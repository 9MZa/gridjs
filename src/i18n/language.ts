import enUS from './en_US';
export type Language = { [key: string]: string | Language };

export class Translator {
  private readonly _language: Language;
  private readonly _defaultLanguage: Language;

  constructor(language?: Language) {
    this._language = language;
    this._defaultLanguage = enUS;
  }

  /**
   * Tries to split the message with "." and find
   * the key in the given language
   *
   * @param message
   * @param lang
   */
  getString(message: string, lang: Language): string {
    if (!lang || !message) return null;

    const splitted = message.split('.');
    const key = splitted[0];

    if (lang[key]) {
      const val = lang[key];

      if (typeof val === 'string') {
        return val;
      } else {
        return this.getString(splitted.slice(1).join('.'), val);
      }
    }

    return null;
  }

  translate(message: string): string {
    const translated = this.getString(message, this._language);

    if (translated) {
      return String(translated);
    }

    return this.getString(message, this._defaultLanguage);
  }
}

export function useTranslator(translator: Translator) {
  return function (message: string): string {
    return translator.translate(message);
  };
}
