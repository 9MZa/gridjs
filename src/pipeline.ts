class Pipeline<T> extends Promise<T>{}
export default Pipeline;
