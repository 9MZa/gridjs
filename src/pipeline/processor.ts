export enum ProcessorType {
  Extractor,
  Transformer,
  Search,
  Sort,
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface PipelineProcessorProps {}

export abstract class PipelineProcessor<T, P extends PipelineProcessorProps> {
  private readonly _props: P;
  private propsUpdatedCallback: Set<(...args) => void> = new Set();

  abstract get type(): ProcessorType;
  abstract process(...args): T | Promise<T>;

  constructor(props?: P) {
    this._props = {} as P;

    if (props) this.setProps(props);
  }

  private trigger(fns: Set<(...args) => void>, ...args): void {
    if (fns) {
      fns.forEach(fn => fn(...args));
    }
  }

  setProps(props: P): this {
    Object.assign(this._props, props);
    this.trigger(this.propsUpdatedCallback);
    return this;
  }

  get props(): P {
    return this._props;
  }

  propsUpdated(callback: (...args) => void): this {
    this.propsUpdatedCallback.add(callback);
    return this;
  }
}
