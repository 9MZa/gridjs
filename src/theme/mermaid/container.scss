.gridjs {
  font-family: Inter var,system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
  font-size: 1rem;
  line-height: 1.25rem;

  &-container {
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    box-shadow: 0 1px 3px 0 rgba(0,0,0,.1), 0 1px 2px 0 rgba(0,0,0,.26);
    border-radius: .5rem;
    overflow: hidden;
    display: inline-block;

    border-bottom-width: 1px;
    border-top-width: 1px;
    border-color: #e5e7eb;
  }
}

