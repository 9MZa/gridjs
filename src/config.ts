import { OneDArray, ProtoExtends, TCell, TColumn, TwoDArray } from './types';
import Storage from './storage/storage';
import Pipeline from './pipeline/pipeline';
import Tabular from './tabular';
import { SearchConfig } from './view/plugin/search/search';
import { PaginationConfig } from './view/plugin/pagination';
import Header from './header';
import { ServerStorageOptions } from './storage/server';
import Dispatcher from './util/dispatcher';
import { GenericSortConfig } from './view/plugin/sort/sort';
import { Language, Translator } from './i18n/language';
import { createRef, RefObject } from 'preact';
import StorageUtils from './storage/storageUtils';
import PipelineUtils from './pipeline/pipelineUtils';

// Config type used internally
export interface Config {
  dispatcher?: Dispatcher<any>;
  /** container element that is used to mount the Grid.js to */
  container?: Element;
  /** gridjs-temp div which is used internally */
  tempRef?: RefObject<HTMLDivElement>;
  data?:
    | TwoDArray<TCell>
    | (() => TwoDArray<TCell>)
    | (() => Promise<TwoDArray<TCell>>);
  server?: ServerStorageOptions;
  header?: Header;
  /** to parse a HTML table and load the data */
  from: HTMLElement;
  storage: Storage<any>;
  pipeline: Pipeline<Tabular>;
  /** to automatically calculate the columns width */
  autoWidth: boolean;
  /** sets the width of the container and table */
  width: string;
  search: SearchConfig;
  pagination: PaginationConfig;
  sort: GenericSortConfig;
  translator: Translator;
}

// Config type used by the consumers
interface UserConfigExtend {
  columns?: OneDArray<TColumn | string>;
  search: SearchConfig | boolean;
  pagination: PaginationConfig | boolean;
  // implicit option to enable the sort plugin globally
  sort: GenericSortConfig | boolean;
  language: Language;
}

export type UserConfig = ProtoExtends<
  Partial<Config>,
  Partial<UserConfigExtend>
>;

export class Config {
  constructor(config?: Config) {
    // FIXME: not sure if this makes sense because Config is a subset of UserConfig
    const updatedConfig = {
      ...Config.defaultConfig(),
      ...config,
    };

    Object.assign(this, updatedConfig);
  }

  update(updatedConfig: Partial<Config>): Config {
    Object.assign(this, updatedConfig || {});
    return this;
  }

  static defaultConfig(): Config {
    return {
      tempRef: createRef(),
      width: '100%',
      autoWidth: true,
    } as Config;
  }

  static fromUserConfig(userConfig: UserConfig): Config {
    const config = new Config(userConfig as Config);

    config.update({
      dispatcher: new Dispatcher<any>(),
      storage: StorageUtils.createFromUserConfig(userConfig),
    });

    config.update({
      pipeline: PipelineUtils.createFromConfig(config),
    });

    // Sort
    if (typeof userConfig.sort === 'boolean' && userConfig.sort) {
      config.update({
        sort: {
          multiColumn: true,
        },
      });
    }

    // Header
    config.update({
      header: Header.fromUserConfig(config),
    });

    // Pagination
    config.update({
      pagination: {
        enabled:
          userConfig.pagination === true ||
          userConfig.pagination instanceof Object,
        ...(userConfig.pagination as PaginationConfig),
      },
    });

    // Search
    config.update({
      search: {
        enabled:
          userConfig.search === true || userConfig.search instanceof Object,
        ...(userConfig.search as SearchConfig),
      },
    });

    // Translator
    config.update({
      translator: new Translator(userConfig.language),
    });

    return config;
  }
}
