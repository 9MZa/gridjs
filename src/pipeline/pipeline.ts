import { PipelineProcessor } from './processor';

class Pipeline<T, P> {
  // available steps for this pipeline
  private readonly _steps: PipelineProcessor<T, P>[] = [];
  private propsUpdatedCallback: Set<(...args) => void> = new Set();
  private updatedCallback: Set<(...args) => void> = new Set();

  constructor(steps?: PipelineProcessor<T, P>[]) {
    if (steps) {
      steps.forEach(step => this.register(step));
    }
  }

  register(processor: PipelineProcessor<T, P>): void {
    if (processor.type === null) {
      throw Error('Processor type is not defined');
    }

    processor.propsUpdated(this.processorPropsUpdated.bind(this));
    this._steps.push(processor);
    this.trigger(this.updatedCallback, processor);
  }

  get steps(): PipelineProcessor<T, P>[] {
    return this._steps;
  }

  process(data: T): T {
    return this._steps.reduce(
      (previous, current) => current.process(previous),
      data,
    );
  }

  private trigger(fns: Set<(...args) => void>, ...args): void {
    if (fns) {
      fns.forEach(fn => fn(...args));
    }
  }

  private processorPropsUpdated(): void {
    this.trigger(this.propsUpdatedCallback);
    this.trigger(this.updatedCallback);
  }

  propsUpdated(fn: (...args) => void): this {
    this.propsUpdatedCallback.add(fn);
    return this;
  }

  updated(fn: (...args) => void): this {
    this.updatedCallback.add(fn);
    return this;
  }
}

export default Pipeline;
