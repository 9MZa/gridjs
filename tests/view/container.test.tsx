import { mount } from 'enzyme';
import { JSDOM } from 'jsdom';
import { h } from 'preact';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Config } from '../../src/config';
import { Container } from '../../src/view/container';
import Pipeline from '../../src/pipeline/pipeline';
import StorageExtractor from '../../src/pipeline/extractor/storage';
import ArrayToTabularTransformer from '../../src/pipeline/transformer/arrayToTabular';
import StorageUtils from '../../src/storage/storageUtils';
import Header from '../../src/header';
import Dispatcher from '../../src/util/dispatcher';
import { Translator } from '../../src/i18n/language';
import { PipelineProcessor, ProcessorType } from '../../src/pipeline/processor';
import * as width from '../../src/util/width';
import { flushPromises } from '../testUtil';

expect.extend(toHaveNoViolations);

describe('Container component', () => {
  let config: Config;

  beforeEach(() => {
    config = new Config();
    config.data = [
      [1, 2, 3],
      ['a', 'b', 'c'],
    ];
    config.autoWidth = true;
    config.storage = StorageUtils.createFromUserConfig(config);
    config.dispatcher = new Dispatcher();
    config.translator = new Translator();
    config.pipeline = new Pipeline([
      new StorageExtractor({ storage: config.storage }),
      new ArrayToTabularTransformer(),
    ]);
  });

  afterAll(() => {
    config = null;
  });

  it('should render a container with table', async () => {
    const container = mount(
      <Container config={config} pipeline={config.pipeline} />,
    );

    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should render a container with searchable table', async () => {
    config.search = {
      enabled: true,
    };

    const container = mount(
      <Container config={config} pipeline={config.pipeline} />,
    );
    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should render a container with sortable and paginated table', async () => {
    config.search = {
      enabled: true,
    };

    config.pagination = {
      enabled: true,
      limit: 5,
    };

    config.header = new Header();
    config.header.columns = [
      {
        name: 'c1',
        sort: {
          enabled: false,
        },
      },
      {
        name: 'c2',
        sort: {
          enabled: true,
        },
      },
      {
        name: 'c3',
        sort: {
          enabled: true,
        },
      },
    ];

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        header={config.header}
      />,
    );
    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should render a container with error', async () => {
    console.error = jest.fn();

    class ErrorProcessor extends PipelineProcessor<string, {}> {
      get type(): ProcessorType {
        return ProcessorType.Limit;
      }
      _process(): string {
        throw Error('something happened');
      }
    }

    config.pipeline.register(new ErrorProcessor());

    const container = mount(
      <Container config={config} pipeline={config.pipeline} />,
    );

    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should not violate accessibility test', async () => {
    config.pagination = {
      enabled: true,
      limit: 1,
    };

    config.search = {
      enabled: true,
    };

    config.sort = {};

    const container = mount(
      <Container config={config} pipeline={config.pipeline} />,
    );

    await container.instance().componentDidMount();
    expect(await axe(container.html())).toHaveNoViolations();
  });

  it('should render a table with correct th widths', async () => {
    const wrapper = new JSDOM('<div style="width: 500px"></div>');

    const mock = jest.spyOn(width, 'getWidth');
    mock.mockReturnValue(42);

    config.container = wrapper;
    config.header = new Header();
    config.header.columns = [
      {
        name: 'hello world',
        sort: {
          enabled: false,
        },
      },
      {
        name: 'you are super cool',
        sort: {
          enabled: true,
        },
      },
      {
        name: 'actions',
        width: '200px',
        sort: {
          enabled: true,
        },
      },
    ];

    const container = mount(
      <Container config={config} pipeline={config.pipeline} />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });
});
