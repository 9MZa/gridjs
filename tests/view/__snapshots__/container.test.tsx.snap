// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component should render a container with error 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\"><div class=\\"gridjs-wrapper\\"><table role=\\"grid\\" class=\\"gridjs-table\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td role=\\"alert\\" colspan=\\"0\\" class=\\"gridjs-td gridjs-message gridjs-error\\">An error happened while fetching the data</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with searchable table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\"><div class=\\"gridjs-head\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\"><table role=\\"grid\\" class=\\"gridjs-table\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with sortable and paginated table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\"><div class=\\"gridjs-head\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\"><table role=\\"grid\\" class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\">c1</th><th class=\\"gridjs-th gridjs-th-sort\\" tabindex=\\"0\\">c2<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th class=\\"gridjs-th gridjs-th-sort\\" tabindex=\\"0\\">c3<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div><div class=\\"gridjs-footer\\"><div class=\\"gridjs-pagination\\"><div role=\\"status\\" class=\\"gridjs-summary\\" title=\\"Page 1 of 1\\">Showing <b>1</b> to <b>2</b> of <b>2</b> results</div><div class=\\"gridjs-pages\\"><button tabindex=\\"0\\" disabled=\\"\\">Previous</button><button tabindex=\\"0\\" class=\\"gridjs-currentPage\\" title=\\"Page 1\\">1</button><button tabindex=\\"0\\" disabled=\\"\\">Next</button></div></div></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\"><div class=\\"gridjs-wrapper\\"><table role=\\"grid\\" class=\\"gridjs-table\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a table with correct th widths 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\"><div class=\\"gridjs-wrapper\\"><table role=\\"grid\\" class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\" style=\\"width: 42px;\\">hello world</th><th class=\\"gridjs-th gridjs-th-sort\\" style=\\"width: 42px;\\" tabindex=\\"0\\">you are super cool<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th class=\\"gridjs-th gridjs-th-sort\\" style=\\"width: 200px;\\" tabindex=\\"0\\">actions<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;
